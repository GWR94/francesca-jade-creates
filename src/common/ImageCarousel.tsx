import React from "react";
import { Carousel, CarouselIndicators, CarouselItem, CarouselControl } from "reactstrap";
import { S3Image } from "aws-amplify-react";
import { API, Storage, graphqlOperation } from "aws-amplify";
import { Button, Dialog } from "@blueprintjs/core";
import { S3ImageProps } from "./interfaces/Product.i";
import { updateProduct } from "../graphql/mutations";
import { Toaster } from "../utils";

interface Props {
  images: S3ImageProps[];
  deleteImages?: boolean;
  id?: string;
  handleUpdateImages?: (image: S3ImageProps[]) => void;
  update?: boolean;
}

interface State {
  animating: boolean;
  currentIndex: number;
  dialogOpen: boolean;
  keyToDelete: string;
}

class ImageCarousel extends React.Component<Props, State> {
  public readonly state: State = {
    animating: false,
    currentIndex: 0,
    dialogOpen: false,
    keyToDelete: null,
  };

  public handleNextImage = (): void => {
    const { animating, currentIndex } = this.state;
    const { images } = this.props;
    if (animating) return;
    this.setState({
      currentIndex: currentIndex === images.length - 1 ? 0 : currentIndex + 1,
    });
  };

  public handlePreviousImage = (): void => {
    const { animating, currentIndex } = this.state;
    const { images } = this.props;
    if (animating) return;
    this.setState({
      currentIndex: currentIndex === 0 ? images.length - 1 : currentIndex - 1,
    });
  };

  public goToIndex = (idx: number): void => {
    const { animating } = this.state;
    if (animating) return;
    this.setState({ currentIndex: idx });
  };

  public handleDeleteImage = async (): Promise<void> => {
    const { keyToDelete } = this.state;
    const { images, update, id, handleUpdateImages } = this.props;
    try {
      const updatedImages = images.filter((img) => img.key !== keyToDelete);
      this.setState({ currentIndex: 0 });
      await Storage.remove(keyToDelete);
      if (update) {
        await API.graphql(
          graphqlOperation(updateProduct, { input: { id, image: updatedImages } }),
        );
      }
      Toaster.show({
        intent: "success",
        message: "Successfully removed image.",
      });
      handleUpdateImages(updatedImages);
      // setDialogOpen(false);
      this.setState({ animating: false });
    } catch (err) {
      console.error(err);
      Toaster.show({
        intent: "danger",
        message: "Failed to removed image. Please try again.",
      });
    }
  };

  public render(): JSX.Element {
    const { images, deleteImages } = this.props;
    const { currentIndex, dialogOpen, keyToDelete } = this.state;
    return (
      <>
        <div className="carousel__container">
          {images.length === 1 ? (
            <div className="update__image-container">
              {deleteImages && (
                <i
                  className="fas fa-times carousel__delete-icon animated infinite pulse"
                  role="button"
                  tabIndex={0}
                  onClick={(): void => {
                    this.setState({
                      dialogOpen: false,
                      keyToDelete: images[0].key,
                    });
                  }}
                />
              )}
              <S3Image
                imgKey={images[0].key}
                theme={{
                  photoImg: {
                    width: "100%",
                  },
                }}
              />
            </div>
          ) : (
            <Carousel
              activeIndex={currentIndex}
              next={this.handleNextImage}
              previous={this.handlePreviousImage}
            >
              <CarouselIndicators
                items={images}
                activeIndex={currentIndex}
                onClickHandler={(idx): void => this.setState({ currentIndex: idx })}
                className="carousel__indicators"
              />
              {images.map((image, i) => (
                <CarouselItem
                  onExiting={(): void => this.setState({ animating: true })}
                  onExited={(): void => this.setState({ animating: false })}
                  key={i}
                >
                  {deleteImages && (
                    <i
                      className="fas fa-times carousel__delete-icon animated infinite pulse"
                      role="button"
                      tabIndex={0}
                      onClick={(): void => {
                        this.setState({
                          dialogOpen: true,
                          keyToDelete: image.key,
                        });
                      }}
                    />
                  )}
                  <div className="update__image-container">
                    <S3Image
                      imgKey={image.key}
                      theme={{
                        photoImg: {
                          width: "100%",
                        },
                      }}
                    />
                  </div>
                </CarouselItem>
              ))}
              <CarouselControl
                direction="prev"
                directionText="Previous"
                onClickHandler={this.handlePreviousImage}
              />
              <CarouselControl
                direction="next"
                directionText="Next"
                onClickHandler={this.handleNextImage}
              />
            </Carousel>
          )}
        </div>
        <Dialog
          isOpen={dialogOpen}
          icon="trash"
          onClose={(): void => this.setState({ dialogOpen: false })}
          title="Are you sure?"
        >
          <div className="update__alert-container">
            <p className="text-center">Are you sure you want to delete this image?</p>
            <p className="text-center">Other images can be added at a later date.</p>

            <S3Image
              imgKey={keyToDelete}
              theme={{
                photoImg: { maxWidth: "100%", marginBottom: "20px" },
              }}
            />
            <div className="dialog__button-container">
              <Button
                intent="danger"
                text="Cancel"
                onClick={(): void => this.setState({ dialogOpen: false })}
                style={{ margin: "0 5px" }}
              />
              <Button
                intent="success"
                text="Confirm"
                onClick={this.handleDeleteImage}
                style={{ margin: "0 5px" }}
              />
            </div>
          </div>
        </Dialog>
      </>
    );
  }
}

export default ImageCarousel;
