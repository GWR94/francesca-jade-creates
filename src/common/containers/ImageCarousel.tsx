import React from "react";
import {
  CarouselProvider,
  Slider,
  Slide,
  ButtonBack,
  ButtonNext,
} from "pure-react-carousel";
import "pure-react-carousel/dist/react-carousel.es.css";
import { S3Image } from "aws-amplify-react";
import { API, Storage, graphqlOperation } from "aws-amplify";
import { ChevronLeft, ChevronRight } from "@material-ui/icons";
import { updateProduct } from "../../graphql/mutations";
import { openSnackbar } from "../../utils/Notifier";
import { ImageCarouselProps, ImageCarouselState } from "./interfaces/ImageCarousel.i";
import { COLORS, INTENT } from "../../themes";
import DeleteImageDialog from "../alerts/DeleteImageDialog";

/**
 * Class component for displaying, updating and removing images for a specific product.
 */
class ImageCarousel extends React.Component<ImageCarouselProps, ImageCarouselState> {
  public readonly state: ImageCarouselState = {
    dialogOpen: false,
    keyToDelete: null,
    isPlaying: true,
  };

  // Ref of carousel so event listeners can be mounted/unmounted to it.
  public carouselRef = React.createRef<HTMLDivElement>();

  /**
   * When the component mounts an event listener should be added to the carousel
   * so it can be paused when hovering and play when removing the mouse from the
   * area - but only when there is 1 or more images in the array.
   */
  public componentDidMount(): void {
    const { images } = this.props;
    if (images.length > 1 && this.carouselRef.current) {
      this.carouselRef.current.addEventListener("mouseenter", (): void => {
        this.setState({ isPlaying: false });
      });
      this.carouselRef.current.addEventListener("mouseleave", (): void => {
        this.setState({ isPlaying: true });
      });
    }
  }

  /**
   * Remove the event listeners when the component unmounts to prevent memory
   * leaks.
   */
  public componentWillUnmount(): void {
    const { images } = this.props;
    if (images.length > 1 && this.carouselRef.current) {
      this.carouselRef.current.removeEventListener("mouseenter", (): void => {
        this.setState({ isPlaying: false });
      });
      this.carouselRef.current.removeEventListener("mouseleave", (): void => {
        this.setState({ isPlaying: true });
      });
    }
  }

  /**
   * Method to remove an image from the carousel, which will inturn remove the
   * image from the database and sever the link between the image and the current
   * product.
   */
  public handleDeleteImage = async (): Promise<void> => {
    // destructure the key to delete!bet wi
    const { keyToDelete } = this.state;
    const { images, update, id, handleUpdateImages } = this.props;
    /**
     * if there is no value in keyToDelete, then the operation cannot be completed,
     * so notify the user of this with a danger snackbar, and return from the function.
     */
    if (!keyToDelete) {
      openSnackbar({
        severity: INTENT.Danger,
        message: "Invalid key to delete. Please try again.",
      });
      return;
    }
    try {
      /**
       * filter out the key (which is saved in state from clicking the remove button
       * on the image in the carousel), so that it's been removed from the array, and
       * therefore updated to its correct data in state.
       */
      const updatedImages = images.filter((img) => img.key !== keyToDelete);
      /**
       * remove the image from the database by providing the key to delete to
       * aws-amplify Storage library
       */
      await Storage.remove(keyToDelete);

      /**
       * If the update prop is set to true, then the link from the image to the product
       * has to be severed, which can be done using the updateProduct graphQL mutation.
       * The previously created updatedImages needs to be passed for the image property
       * so the images left are the correct ones.
       */
      if (update) {
        await API.graphql(
          graphqlOperation(updateProduct, { input: { id, image: updatedImages } }),
        );
      }
      /**
       * Once the operation is complete, let the user know with a success snackbar with
       * a relevant message.
       */
      openSnackbar({
        severity: "success",
        message: "Successfully removed image.",
      });
      // if the handleUpdateImages function is defined, execute it.
      if (handleUpdateImages) handleUpdateImages(updatedImages);
      this.setState({ dialogOpen: false });
    } catch (err) {
      /**
       * If there are any errors at any point in the function, notify the user with a
       * danger snackbar with a relevant message.
       */
      openSnackbar({
        severity: "error",
        message: "Failed to removed image. Please try again.",
      });
    }
  };

  public render(): JSX.Element {
    const { images, deleteImages, type } = this.props;
    const { dialogOpen, keyToDelete, isPlaying } = this.state;
    return (
      <>
        {/* 
          If there is only one item in the images array, then there is no need to render
          the carousel component, so just place the image in a div.
        */}
        {images.length === 1 ? (
          <div style={{ position: "relative" }}>
            {deleteImages && (
              /**
               * if the user is allowed to delete images (is an admin), show a pulsing
               * delete icon which shows the user that they can delete it
               */
              <i
                className="fas fa-times carousel__delete-icon animated infinite pulse"
                role="button"
                tabIndex={0}
                onClick={(): void => {
                  this.setState({
                    dialogOpen: true,
                    keyToDelete: images[0].key,
                  });
                }}
              />
            )}
            <S3Image
              imgKey={images[0].key}
              theme={{
                photoImg: {
                  width: "100%",
                },
              }}
            />
          </div>
        ) : (
          /**
           * If there is more than one image in the array then the carousel needs to be
           * rendered with each of the images mapped to a Slide component.
           */
          <div ref={this.carouselRef} style={{ position: "relative" }}>
            <CarouselProvider
              naturalSlideHeight={6}
              naturalSlideWidth={5}
              totalSlides={images.length}
              visibleSlides={1}
              step={1}
              isPlaying={isPlaying && !dialogOpen}
              infinite
            >
              <Slider>
                {images.map((image, i) => (
                  <Slide index={i} key={i} style={{ position: "relative" }}>
                    {deleteImages && (
                      <i
                        className="fas fa-times carousel__delete-icon animated infinite pulse"
                        role="button"
                        tabIndex={0}
                        onClick={(): void => {
                          this.setState({
                            dialogOpen: true,
                            keyToDelete: image.key,
                          });
                        }}
                      />
                    )}
                    <S3Image
                      imgKey={image.key}
                      theme={{
                        photoImg: {
                          width: "100%",
                        },
                      }}
                    />
                  </Slide>
                ))}
              </Slider>
              {/*  Buttons to control the playback of the carousel */}
              <ButtonBack className="carousel__back animated pulse infinite">
                <ChevronLeft
                  className="carousel__svg"
                  style={{ color: type === "Cake" ? COLORS.Pink : COLORS.Purple }}
                />
              </ButtonBack>
              <ButtonNext className="carousel__forward animated pulse infinite">
                <ChevronRight
                  className="carousel__svg"
                  style={{ color: type === "Cake" ? COLORS.Pink : COLORS.Purple }}
                />
              </ButtonNext>
            </CarouselProvider>
          </div>
        )}
        {/* 
          Dialog to confirm if the user wants to delete the image. Only opens when
          the user clicks the delete icon on the image. 
        */}
        {keyToDelete && (
          <DeleteImageDialog
            dialogOpen={dialogOpen}
            closeDialog={(): void => this.setState({ dialogOpen: false })}
            keyToDelete={keyToDelete}
            handleDeleteImage={this.handleDeleteImage}
          />
        )}
      </>
    );
  }
}

export default ImageCarousel;
