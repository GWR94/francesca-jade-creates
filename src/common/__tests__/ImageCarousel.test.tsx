import React from "react";
import { shallow } from "enzyme";
import { Carousel, CarouselIndicators, CarouselControl } from "reactstrap";
import { S3Image } from "aws-amplify-react";
import ImageCarousel from "../ImageCarousel";

describe("<ImageCarousel />", () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  // @ts-ignore
  useStateSpy.mockImplementation((init) => [init, setState]);

  afterEach(() => {
    jest.clearAllMocks();
  });

  const multipleImages = [
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106247322-IMG_0088.JPG",
    },
  ];

  const images = [
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
    },
  ];

  it("should render correctly when only one image is in image array prop", () => {
    const wrapper = shallow(<ImageCarousel id="test123" images={images} />);
    expect(wrapper).toMatchSnapshot();
    // checks the result is the image container rather than carousel
    expect(
      wrapper.find(".carousel__container").at(0).childAt(0).prop("className"),
    ).toEqual("update__image-container");
    // check the child is of type S3Image rather than carousel item.
    expect(wrapper.find(".update__image-container").at(0).childAt(0).type()).toBe(
      S3Image,
    );
  });

  it("should render correctly when multiple images in image array prop", () => {
    const wrapper = shallow(<ImageCarousel id="test321" images={multipleImages} />);
    expect(wrapper).toMatchSnapshot();
    const carousel = wrapper.find(".carousel__container").at(0).childAt(0);
    expect(carousel.type()).toBe(Carousel);
    expect(carousel.childAt(0).type()).toBe(CarouselIndicators);
    const image = wrapper.find(".update__image-container").at(0).childAt(0);
    expect(image.type()).toEqual(S3Image);
    expect(image.prop("imgKey")).toBe(multipleImages[0].key);
  });

  it("should change currentIndex state which controls what image is shown when clicking on the next carousel button", () => {
    const wrapper = shallow(<ImageCarousel id="test123" images={multipleImages} />);
    const image = wrapper.find(".update__image-container").at(0).childAt(0);
    expect(image.type()).toBe(S3Image);
    // check initial state
    expect(wrapper.find(Carousel).prop("activeIndex")).toBe(0);
    expect(wrapper.state("currentIndex")).toBe(0);
    // click next control button
    wrapper.find(CarouselControl).at(1).props().onClickHandler();
    expect(wrapper.state("currentIndex")).toBe(1);
    expect(wrapper.find(Carousel).at(0).prop("activeIndex")).toBe(1);
    // click next to wrap round to start of the images (0)
    wrapper.find(CarouselControl).at(1).props().onClickHandler();
    expect(wrapper.state("currentIndex")).toBe(0);
    expect(wrapper.find(Carousel).at(0).prop("activeIndex")).toBe(0);
  });

  it("should change currentIndex state which controls what image is shown when clicking prev carousel button", () => {
    const wrapper = shallow(<ImageCarousel id="123test" images={multipleImages} />);
    expect(wrapper.state("currentIndex")).toBe(0);
    // go to the next image
    wrapper.find(CarouselControl).at(1).props().onClickHandler();
    expect(wrapper.state("currentIndex")).toBe(1);
    expect(wrapper.find(Carousel).at(0).prop("activeIndex")).toBe(1);
    // go to previous
    wrapper.find(CarouselControl).at(0).props().onClickHandler();
    expect(wrapper.state("currentIndex")).toBe(0);
    expect(wrapper.find(Carousel).at(0).prop("activeIndex")).toBe(0);
    // previous goes to last image in array (1)
    wrapper.find(CarouselControl).at(0).props().onClickHandler();
    expect(wrapper.state("currentIndex")).toBe(1);
    expect(wrapper.find(Carousel).at(0).prop("activeIndex")).toBe(1);
  });
});
