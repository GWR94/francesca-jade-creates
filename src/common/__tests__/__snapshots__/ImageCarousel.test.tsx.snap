// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageCarousel /> should render correctly when multiple images in image array prop 1`] = `
<Fragment>
  <div
    className="carousel__container"
  >
    <Carousel
      activeIndex={0}
      enableTouch={true}
      interval={5000}
      keyboard={true}
      next={[Function]}
      pause="hover"
      previous={[Function]}
      slide={true}
    >
      <CarouselIndicators
        activeIndex={0}
        className="carousel__indicators"
        items={
          Array [
            Object {
              "bucket": "fjadecreatesmedia121747-production",
              "key": "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
              "region": "eu-west-2",
            },
            Object {
              "bucket": "fjadecreatesmedia121747-production",
              "key": "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106247322-IMG_0088.JPG",
              "region": "eu-west-2",
            },
          ]
        }
        onClickHandler={[Function]}
      />
      <CarouselItem
        appear={false}
        enter={true}
        exit={true}
        in={false}
        key="0"
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        slide={true}
        tag="div"
        timeout={600}
        unmountOnExit={false}
      >
        <div
          className="update__image-container"
        >
          <S3Image
            imgKey="/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG"
            theme={
              Object {
                "photoImg": Object {
                  "width": "100%",
                },
              }
            }
          />
        </div>
      </CarouselItem>
      <CarouselItem
        appear={false}
        enter={true}
        exit={true}
        in={false}
        key="1"
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        slide={true}
        tag="div"
        timeout={600}
        unmountOnExit={false}
      >
        <div
          className="update__image-container"
        >
          <S3Image
            imgKey="/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106247322-IMG_0088.JPG"
            theme={
              Object {
                "photoImg": Object {
                  "width": "100%",
                },
              }
            }
          />
        </div>
      </CarouselItem>
      <CarouselControl
        direction="prev"
        directionText="Previous"
        onClickHandler={[Function]}
      />
      <CarouselControl
        direction="next"
        directionText="Next"
        onClickHandler={[Function]}
      />
    </Carousel>
  </div>
  <Blueprint3.Dialog
    canOutsideClickClose={true}
    icon="trash"
    isOpen={false}
    onClose={[Function]}
    title="Are you sure?"
  >
    <div
      className="update__alert-container"
    >
      <p
        className="text-center"
      >
        Are you sure you want to delete this image?
      </p>
      <p
        className="text-center"
      >
        Other images can be added at a later date.
      </p>
      <S3Image
        imgKey={null}
        theme={
          Object {
            "photoImg": Object {
              "marginBottom": "20px",
              "maxWidth": "100%",
            },
          }
        }
      />
      <div
        className="dialog__button-container"
      >
        <Blueprint3.Button
          intent="danger"
          onClick={[Function]}
          style={
            Object {
              "margin": "0 5px",
            }
          }
          text="Cancel"
        />
        <Blueprint3.Button
          intent="success"
          onClick={[Function]}
          style={
            Object {
              "margin": "0 5px",
            }
          }
          text="Confirm"
        />
      </div>
    </div>
  </Blueprint3.Dialog>
</Fragment>
`;

exports[`<ImageCarousel /> should render correctly when only one image is in image array prop 1`] = `
<Fragment>
  <div
    className="carousel__container"
  >
    <div
      className="update__image-container"
    >
      <S3Image
        imgKey="/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG"
        theme={
          Object {
            "photoImg": Object {
              "width": "100%",
            },
          }
        }
      />
    </div>
  </div>
  <Blueprint3.Dialog
    canOutsideClickClose={true}
    icon="trash"
    isOpen={false}
    onClose={[Function]}
    title="Are you sure?"
  >
    <div
      className="update__alert-container"
    >
      <p
        className="text-center"
      >
        Are you sure you want to delete this image?
      </p>
      <p
        className="text-center"
      >
        Other images can be added at a later date.
      </p>
      <S3Image
        imgKey={null}
        theme={
          Object {
            "photoImg": Object {
              "marginBottom": "20px",
              "maxWidth": "100%",
            },
          }
        }
      />
      <div
        className="dialog__button-container"
      >
        <Blueprint3.Button
          intent="danger"
          onClick={[Function]}
          style={
            Object {
              "margin": "0 5px",
            }
          }
          text="Cancel"
        />
        <Blueprint3.Button
          intent="success"
          onClick={[Function]}
          style={
            Object {
              "margin": "0 5px",
            }
          }
          text="Confirm"
        />
      </div>
    </div>
  </Blueprint3.Dialog>
</Fragment>
`;
